### DISCUSSION TITLE
Laravel 10.x Model Attribute Mutators support for both snake_case and camelCase Columns.

### DISCUSSION THREAD

**Discussion by jcrawford** [laravel/framework#48585] [Ideas]
### Laravel Version

10.8

### PHP Version

8.2.10

### Database Driver & Version

MySQL on AWS

### Description

I am trying to use Laravel 10.8 mutators using the Attribute return type.  The issue I am having is that some of our database columns are `snake_case` and some are `camelCase`.  Having both column types is something that we currently have to live with as modifying the table with millions of records is not an option at this time.  What I am finding is that when I call `$model->toArray()` the mutators are being executed for our `snake_case` columns but not for our `camelCase` columns.  

It appears to be failing in the `HasAttributes` trait [on the following lines](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L228-L230).  It seems to think the `camelCase` attribute is `snake_case` and since it does not find the snake case attribute it `continue`s.

I have found that for some reason in the HasAttributes.php file our column of `camelCaseName` is listed in the `static::$snakeAttributes` array on the model as `snake_case_name`, therefore it thinks there is a column `snake_case_name` on our model so I assume that the mutator is trying to mutate that attribute which does not exist.  Currently all of our `camelCase` attributes are ignored when the mutator is executed when the model is converted to an Array.  

The `camelCase` mutators only seem to work if we manually call `$model->camelCaseName`.  In this case the mutator is executed.  I have tried setting the models `$snakeAttributes` property to `false` as it is true by default, when I do this it causes issues with our mutators for `snake_case` columns.  Is there a way to make Laravel support both snake and camel case attributes when using mutators?

I have found a work-around but it is tedious and requires all developers to `remember ` they have to do this, we would have to set `$snakeAttributes = false` which will allow the `snakeCase` column and mutator to run, however if we then add mutators for our `snake_case` columns we have to define two methods for it to work.  If we define only the `snake_case` mutator then an error is thrown that it cannot find the `camelCaseName()` method.  If we don't define the `camelCaseName` mutator method then the mutation is not completed on the `snake_case` attribute.

In order to make this work for both camelCase and snake_case columns on a model where both will have Attribute mutators you need to set `$snakeAttributes = false` and then for all `snake_case` columns that have mutators you will need to do the following

```php
use Illuminate\Database\Eloquent\Model;

class MyModel extends Model {

  public static $snakeAttributes = false;
  
  /**
   * This is the mutator for the camelCase column
   */
  public function camelCaseColumnName() : Attribute
  {
    return Attribute::make(
      get: fn (float $value) => intval($value * 100),
      set: fn (int $value) => $value / 100.0
    );
  }

 /**
  * This has to be here otherwise the mutation does not happen.
  */
  public function snake_case_column_name(): Attribute
  {
    return $this->snakeCaseColumnName();
  }

  /**
   * This is the camel case mutator for the snake_case_column_name column
   */
  public function snakeCaseColumnName(): Attribute
  {
    return Attribute::make(
      get: fn (float $value) => intval($value * 100),
      set: fn (int $value) => $value / 100.0
    );
  }
}
```

I have been trying to trace the issue through HasAttributes but was unable to determine why this is happening and ran out of time at the moment.

### Steps To Reproduce

Create a database with a table having both snake_case and camelCase columns of type Decimal.
Create a model for the table using the code in the description naming your mutators to match your column names.
Add the following model property `public static $snakeAttributes = false;`
Start Tinker
Find a record and assign to a variable `$variable = MyModel::find(1);`
Call `$variable->toArray()` and notice the outcome is as explained in the description.

Remove the `snake_case` mutator from your model
Restart Tinker for code changes.
Find a record and assign to a varable `$variable = MyModel::find(1);`
Call `$variable->toArray()` notice there are no errors but the `snake_case` mutator did not mutate the data.

Remove the `camelCase` mutator for the `snake_case` column from your model
Restart Tinker for code changes.
Find a Find a record and assign to a varable `$variable = MyModel::find(1);`
Call `$variable->toArray()` notice the erorr that `camelCaseName` method was not defined.

Remove the `$snakeAttributes = false` from the model and then
Restart Tinker for code changes.
Find a record and assign to a variabel `$variable = MyModel::find(1);`
call `$variable->toArray()` and notice the `camelCase` column value is not mutated but the `snake_case` value was.

**Comment by Eforen** [comment]
Nice catch

**Comment by josephcrawfordSRH** [comment]
Why was this changed to a discussion?  This seems to be a legitimate issue with Eloquent when it comes to `snake_case` and `camelCase` columns being used at the same time.  The accessors work perfectly fine when you directly access the column on the model, they only start acting strange when `toArray()` is being called.

The documentation for mutators/accessors state that you should use the `camelCase` version of the column name when applicable.

[Defining An Accessor](https://laravel.com/docs/9.x/eloquent-mutators#defining-an-accessor)
An accessor transforms an Eloquent attribute value when it is accessed. To define an accessor, create a protected method on your model to represent the accessible attribute. This method name should correspond to the "camel case" representation of the true underlying model attribute / database column when applicable.

In our case the columns is `camelCaseName` and the mutator/accessor is `camelCaseName`.  When the [following code is executed](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L2198) it is adding the  column `camel_case_name` to the list of mutator columns since by default `$snakeAttributes` is set to true.  

Once the code gets to[ this point in the execution](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L253-L255) it says there is no such column `camel_case_name` so we will just continue on and ignore the mutator/accessor modifications to the column.  It is true there is no `camel_case_name` column in the model attributes array since it is `camelCaseName`.

Wouldn't it just make more sense in[ the following method ](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L2191-L2206)to first check for the snake case attribute and then if not found check for the camel case attribute?

I am honestly not sure why this is such a large issue either, when I search the Illuminate folder for the string `$snakeAttributes` it is only found in this trait in 3 places outside of the property definition, two of which are in the `cacheMutatedAttributes` method.

1. [relationsToArray](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L376-L378)
2. [cacheMutatedAttributes](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L2195-L2205)

It seems like a check could be made on the `attributes` to see if the column is camel or snake case before just assuming it is and applying based on a property value that can be set by a developer.

**Reply by Angel5a** [reply]
I think, because having both snake cased and camel cased columns at the same time is a "bad practice". So, should it really be handled by the core framework, with a penalty for a "proper setup" is a question.

**Reply by josephcrawfordSRH** [reply]
Yea I get that but why not simply modify [this block of code](https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L2195-L2205) to not check the snakeAttributes and instead check the $attributes for both versions?  Why do mutators require snake_case but other features do not?  It seems to only be in use in the attributes trait.
